### PR TITLE
Add design_library module to be shared across app

### PR DESCRIPTION
**Pull request by ataulm**
This PR adds some UI for the Actor Detail screen. 

We extract a custom view with the intention of using it in the app. (image + title) The view encapsulates layout logic only, and exposes its own model to highlight valid configurations.

This PR also collapses the `resources` module into `design_library`.

Feed | Actor detail
---|---
feed | actordetail

### CODE REVIEW

File: actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailFilmAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import android.view.View
+import android.view.ViewGroup
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import com.muvi.design_library.ListItem1View
+import com.muvi.design_library.commaAmpersandList
+import com.muvi.design_library.R as designLibrary
+
+internal class ActorDetailFilmAdapter : ListAdapter<UiModel.Film, ActorDetailFilmAdapter.ViewHolder>(Differ) {
```
Comment (ataulm) [inline actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailFilmAdapter.kt]: This adapter currently has 1 view type, the `ListItem1View` which is also used in `feed`. Eventually it'll be nice to include information about the actor themselves at the top (a new viewtype).

File: actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailFilmAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        val uiModel = getItem(position)
+        val model = ListItem1View.Model(
+                poster = uiModel.poster,
+                title = title(uiModel.title, uiModel.characterName, uiModel.year, uiModel.directors),
+                onClick = uiModel.onClick
+        )
+        (holder.itemView as ListItem1View).bind(model)
+    }
+
+    private fun title(title: String, characterName: CharSequence?, year: CharSequence?, directors: List<String>): CharSequence {
```
Comment (ataulm) [inline actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailFilmAdapter.kt]: The `feed` one doesn't include character name. We're using the same view (ListItem1) so it'll look the same style/consistent, but binding different properties.

File: actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt [hunk 3]
```diff
@@ -16,20 +17,53 @@ internal class ActorDetailViewModel(
...
+    )
+
+    sealed class Event {
+        data class NavigateToFilmDetail(val filmId: String) : Event()
+    }
+}
+
+internal data class UiModel(
+        // TODO: we want more actor info here (name, image, description)
+        val films: List<Film>
```
Comment (ataulm) [inline actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt]: This is used directly by the adapter. If we included the actor info, would we include them "ready to use" by the adapter?

i.e.:

```kotlin
internal data class UiModel(val items: List<Item>) {
    sealed class Item {
        data class Header(...) : Item()
        data class Film(...) : Item()
    }
}
```

vs

```kotlin
internal data class UiModel(val header:Header, val films: List<Film>) {
        data class Header(...)
        data class Film(...)
}
```

I think the former is easier but maybe the latter is more correct (and the activity/fragment is responsible for mapping the `UiModel` to the adapter items)

File: base_data/src/main/java/com/muvi/base_data/DbFilmSummary.kt [hunk 4]
```diff
@@ -9,6 +9,9 @@ data class DbFilmSummary(
         @ColumnInfo(name = "release_year") val releaseYear: Int?,
         @field:TypeConverters(StringListConverters::class)
         @ColumnInfo(name = "directors") val directors: List<String>,
+        // TODO: It's not safe to serialise models like this because
+        //  the schema can change — we'd have to nuke the DB on upgrade (sure)
+        //  or provide migrations between versions (NO)
         @Embedded val poster: DbImage
```
Comment (ataulm) [inline base_data/src/main/java/com/muvi/base_data/DbFilmSummary.kt]: Not a big deal for this demo, but not great for real life project

File: design_library/build.gradle [hunk 5]
```diff
@@ -0,0 +1,15 @@
+apply plugin: 'com.android.library'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-android-extensions'
+
+dependencies {
+    // Not sure if this is a bug or expected, but anyone using design_library needs constraintLayout also, hence `api`
+    // e: Supertypes of the following classes cannot be resolved. Please make sure you have the required dependencies in the classpath:
+    // class com.muvi.design_library.ListItem1View, unresolved supertypes: androidx.constraintlayout.widget.ConstraintLayout
+    api libraries.androidxConstraintLayout
```
Comment (ataulm) [inline design_library/build.gradle]: `actor_detail` doesn't use ConstraintLayout directly - only in the inflation of `ListItem1View`.

Expectation was that it wouldn't need `ConstraintLayout` on the class path (it's compilation error). I wonder if adding `compileOnly ConstraintLayout` would work for `actor_detail` but don't want to add that everywhere.

File: design_library/build.gradle [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+    implementation libraries.androidxAppCompat
```
Comment (ataulm) [inline design_library/build.gradle]: @ricknout is there a more explicit dependency we could use here? In my understanding (maybe incorrect) `androidxAppCompat` is a grouping of other dependencies (like the fragment one). 

I think we need this here for AppCompatActivity for the playground.

File: design_library/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest package="com.muvi.design_library"
+    xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <application>
```
Comment (ataulm) [inline design_library/src/main/AndroidManifest.xml]: Just moved from `resources`

File: design_library/src/main/java/com/muvi/design_library/ListItem1View.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package com.muvi.design_library
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.View
+import androidx.constraintlayout.widget.ConstraintLayout
+import com.bumptech.glide.Glide
+import kotlinx.android.synthetic.main.merge_list_item_1.view.*
+
+class ListItem1View(context: Context, attrs: AttributeSet) : ConstraintLayout(context, attrs) {
```
Comment (ataulm) [inline design_library/src/main/java/com/muvi/design_library/ListItem1View.kt]: `ListItem1` is maybe a little redundant until we have `ListItem2` but I wanted to highlight that this is part of a design system with symbols and naming from the design team.

File: design_library/src/main/res/layout/item_film_summary.xml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<com.muvi.design_library.ListItem1View xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (ataulm) [inline design_library/src/main/res/layout/item_film_summary.xml]: This layout is in `design_library` but I dunno if there'd be cases where we'd need to adjust the layout attributes for a specific usecase (in one module but not the others).

File: feed/build.gradle [hunk 10]
```diff
@@ -7,17 +7,20 @@ dependencies {
...
     implementation project(':core')
+    implementation project(':design_library')
     implementation project(':feed_cache')
     implementation project(':feed_data')
     implementation project(':feed_domain')
     implementation project(':feed_remote')
     implementation project(':navigation')
 
     implementation libraries.androidxAppCompat
+    implementation libraries.androidxConstraintLayout
```
Comment (ataulm) [inline feed/build.gradle]: I don't think this is necessary 🤔 I probably started with ListItem1View in this repo and then moved it out.

File: feed/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -1,13 +1,2 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:dist="http://schemas.android.com/apk/distribution"
-    package="com.muvi.feed">
-
-    <dist:module
-        dist:instant="false"
-        dist:onDemand="false"
-        dist:title="@string/title_film_detail">
-        <dist:fusing dist:include="true" />
```
Comment (ataulm) [inline feed/src/main/AndroidManifest.xml]: bad copy paste

File: actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt (hunk 3 above)
Comment (hitherejoe) [reply actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt]: Yea I don't think the adapter should be responsible for manipulating this data in any way, having it ready to use would be nicer

File: actor_detail/src/main/res/layout/activity_actor_detail.xml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.recyclerview.widget.RecyclerView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/recyclerView"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager" />
```
Comment (ricknout) [inline actor_detail/src/main/res/layout/activity_actor_detail.xml]: FWIW, if all of the RV's in the app are using `LinearLayoutManager`, you can now generalise this with a new `recyclerViewStyle` theme attr (https://developer.android.com/jetpack/androidx/releases/recyclerview#1.1.0-beta02):

In `styles.xml`
```
<style name="Widget.Muvi.RecyclerView" parent="">
    <item name="layoutManager">androidx.recyclerview.widget.LinearLayoutManager</item>
</style>
```
In `themes.xml`:
```
<style name="Theme.Muvi" parent="Theme.MaterialComponents.DayNight.DarkActionBar">
    ...
    <item name="recyclerViewStyle">@style/Widget.Muvi.RecyclerView</item>
</style>

<style name="Theme.Muvi.NoActionBar" parent="Theme.MaterialComponents.DayNight.NoActionBar">
    ...
    <item name="recyclerViewStyle">@style/Widget.Muvi.RecyclerView</item>
</style>
```

File: actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt (hunk 3 above)
Comment (ataulm) [reply actor_detail/src/main/java/com/muvi/actor_detail/ActorDetailViewModel.kt]: works for me

File: actor_detail/src/main/res/layout/activity_actor_detail.xml (hunk 12 above)
Comment (ataulm) [reply actor_detail/src/main/res/layout/activity_actor_detail.xml]: :hero:

**Pull request merged**